pkg/build/strategies/sti/sti.go

```diff
@@ -620,11 +646,295 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
+	}
+	defer func() {
+		fd.Close()
+		os.Remove(fd.Name())
+	}()
+
+	if err := builder.docker.DownloadFromContainer(artifactPath, fd, containerID); err != nil {
+		return fmt.Errorf("Couldn't download file (%q -> %q) from container %s: %v", artifactPath, fd.Name(), containerID, err)
+	}
+
+	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
+	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
+		return fmt.Errorf("Couldn't seek to the beginning of the file %q: %v", fd.Name(), err)
+	}
+
+	if err := builder.tar.ExtractTarStream(artifactsDir, fd); err != nil {
+		return fmt.Errorf("Couldn't extract runtime artifact %q into the directory %q: %v", artifactPath, artifactsDir, err)
+	}
+
+	return nil
+}
+
+func (builder *STI) downloadFilesFromBuilderImageStep(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing post execute step: download files from the builder image")
+
+	artifactsDir := filepath.Join(builder.config.WorkingDir, api.RuntimeArtifactsDir)
+	if err := builder.fs.Mkdir(artifactsDir); err != nil {
+		return fmt.Errorf("Couldn't create directory %q: %v", artifactsDir, err)
+	}
+
+	for _, artifact := range builder.config.RuntimeArtifacts {
+		if err := builder.downloadAndExtractFile(artifact.Source, artifactsDir, ctx.containerID); err != nil {
+			return err
+		}
+
+		// for mapping like "/tmp/foo.txt -> app" we should create "app" and move "foo.txt" to that directory
+		dstSubDir := path.Clean(artifact.Destination)
+		if dstSubDir != "." && dstSubDir != "/" {
+			dstDir := filepath.Join(artifactsDir, dstSubDir)
+			glog.V(2).Infof("Creating directory %q", dstDir)
+			if err := os.MkdirAll(dstDir, 0755); err != nil {
+				return fmt.Errorf("Couldn't create directory %q: %v", dstDir, err)
+			}
+
+			file := filepath.Base(artifact.Source)
+			old := filepath.Join(artifactsDir, file)
+			new := filepath.Join(artifactsDir, dstSubDir, file)
+			glog.V(2).Infof("Renaming %q to %q", old, new)
+			if err := os.Rename(old, new); err != nil {
+				return fmt.Errorf("Couldn't rename %q -> %q: %v", old, new, err)
+			}
+		}
+
+	}
+
+	// We copy run script to the directory with artifacts to upload files in one shot
+	runScriptSrcPath := filepath.Join(builder.config.WorkingDir, api.UploadScripts, api.Run)
+	runScriptDstPath := filepath.Join(artifactsDir, api.Run)
+	if err := builder.fs.Copy(runScriptSrcPath, runScriptDstPath); err != nil {
+		return fmt.Errorf("Couldn't copy file (%q -> %q): %v", runScriptSrcPath, runScriptDstPath, err)
+	}
+
+	return nil
+}
+
+func (builder *STI) startRuntimeImageAndUploadFilesStep(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing post execute step: start runtime image and upload files")
+
+	outReader, outWriter := io.Pipe()
+	defer outReader.Close()
+	defer outWriter.Close()
+
+	errReader, errWriter := io.Pipe()
+	defer errReader.Close()
+	defer errWriter.Close()
+
+	lastFileDstPath := "/tmp/upload-done"
```

Comment:
> nm, i guess the ioutil.TempFile below will append a unique suffix?

Yes, but this file is the source file that will be copied into the container to the file with "static" name:
```
I0609 14:57:44.790619 03525 sti.go:877] Uploading file "/tmp/s2i-upload-done559194078" -> "/tmp/upload-done"
```
I'll randomize the name.